### PR TITLE
Add AddHttpClientDefaults

### LINKED ISSUE #87914: [API Proposal] ConfigureHttpClientDefaults for HttpClientFactory
<details>
<summary>Original issue by @JamesNK</summary>

### Background and motivation

`HttpClientFactory` in Microsoft.Extensions.Http allows a developer to centrally configure and instatiate HTTP clients with DI:

```csharp
services
    .AddHttpClient("consoto", c => c.BaseAddress = new Uri("https://consoto.com/"))
    .AddHttpMessageHandler<MyAuthHandler>();

services
    .AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"))
    .AddHttpMessageHandler<MyAuthHandler>();
```

Two named clients are defined in the example above. Unfortunately, there isn't a way to apply the default settings for all clients. Both need `AddHttpMessageHandler<MyAuthHandler>()`.

This issue proposes an `AddHttpClientDefaults` method that can be used to specify configuration that is applied to all clients.

### API Proposal

```diff
namespace Microsoft.Extensions.DependencyInjection;

public static class HttpClientFactoryServiceCollectionExtensions
{
+   public static IHttpClientBuilder AddHttpClientDefaults(this IServiceCollection services);
}
```

### API Usage

`AddHttpClientDefaults` returns an `IHttpClientBuilder`. This is the same type returned by `services.AddHttpClient(...)`. That means all the extension methods for `IHttpClientBuilder` automatically work with `AddHttpClientDefaults`.

```csharp
services.AddHttpClientDefaults()
    .AddHttpMessageHandler<MyAuthHandler>();

// Clients automatically have the handler specified as a default.
services.AddHttpClient("consoto", c => c.BaseAddress = new Uri("https://consoto.com/"));
services.AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"));
```

The default configuration is run on a client before client-specific configuration:

```csharp
services.AddHttpClientDefaults()
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://consoto.com/"));

// Client a base address of https://consoto.com
services.AddHttpClient("consoto");

// Client has a base address of https://github.com/ (overrides default)
services.AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"));
```

Default configuration can be added to multiple times, and its order compared to `AddHttpClient` doesn't matter:

```csharp
services.AddHttpClientDefaults()
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://consoto.com/"));

// Client a base address of https://consoto.com and has MyAuthHandler
services.AddHttpClient("consoto");

services.AddHttpClientDefaults()
    .AddHttpMessageHandler<MyAuthHandler>();
```

### Alternative Designs

We want `AddHttpClientDefaults` that returns an `IHttpClientBuilder` so existing extension methods automatically work with it.

We want to _**avoid**_ the situation that we are seeing where people are adding two extension methods: one to add configuration to a builder, and another to apply configuration to all builders.

Example of what we don't want:

```csharp
// Add handler to one client
services.AddHttpClient("myclient").AddMyCustomerLogging();

// Add handler to all clients
services.AddMyCustomerLoggingToAllClients();
```

### Risks

_No response_

</details>

--------------------

### Background and motivation

`HttpClientFactory` in Microsoft.Extensions.Http allows a developer to configure and create HttpClient instances with DI.

Consider an example where two named clients are defined, and they all need MyAuthHandler in their message handlers chain. 

```csharp
services
    .AddHttpClient("consoto", c => c.BaseAddress = new Uri("https://consoto.com/"))
    .AddHttpMessageHandler<MyAuthHandler>();

services
    .AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"))
    .AddHttpMessageHandler<MyAuthHandler>();
```

Unfortunately, there isn't a way to apply the default settings for all clients. All need `AddHttpMessageHandler<MyAuthHandler>()`.

This issue proposes an `ConfigureHttpClientDefaults` method that can be used to specify configuration that is applied to all clients.

### API Proposal

```c#
// existing
public static class HttpClientFactoryServiceCollectionExtensions
{
    // new
    public static IServiceCollection ConfigureHttpClientDefaults(
        this IServiceCollection services, Action<IHttpClientBuilder> configure) {}
}

// existing
public static class HttpClientBuilderExtensions
{
    // new
    public static IHttpClientBuilder ConfigurePrimaryHttpMessageHandler(
        this IHttpClientBuilder builder,
        Action<HttpMessageHandler, IServiceProvider> configureHandler) {}

    public static IHttpClientBuilder ConfigureAdditionalHttpMessageHandlers(
        this IHttpClientBuilder builder,
        Action<IList<DelegatingHandler>, IServiceProvider> configureAdditionalHandlers) {}


    // existing (+2 overloads)
    // public static IHttpClientBuilder ConfigurePrimaryHttpMessageHandler(this IHttpClientBuilder builder, Func<HttpMessageHandler> configureHandler) {}

    // existing (+2 overloads)
    // public static IHttpClientBuilder AddHttpMessageHandler(this IHttpClientBuilder builder, Func<DelegatingHandler> configureHandler) {}

    // existing -- to be deprecated in the future
    // public static IHttpClientBuilder ConfigureHttpMessageHandlerBuilder(this IHttpClientBuilder builder, Action<HttpMessageHandlerBuilder> configureBuilder) {}
}
```

#### 1. Basic usage and extension methods

`ConfigureHttpClientDefaults` accepts an action over `IHttpClientBuilder`. This is the same type returned by `services.AddHttpClient(...)`. That means all the extension methods for `IHttpClientBuilder` automatically work with `AddHttpClientDefaults`.

```csharp
services.ConfigureHttpClientDefaults(b =>
    b.AddHttpMessageHandler<MyAuthHandler>());

// Clients automatically have the handler specified as a default.
services.AddHttpClient("consoto", c => c.BaseAddress = new Uri("https://consoto.com/"));
services.AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"));
```

#### 2. Order of applying

The default configuration is run on a client before client-specific configuration:

```csharp
services.ConfigureHttpClientDefaults(b =>
    b.ConfigureHttpClient(c => c.BaseAddress = new Uri("https://consoto.com/")));

// Client a base address of https://consoto.com
services.AddHttpClient("consoto");

// Client has a base address of https://github.com/ (overrides default)
services.AddHttpClient("github", c => c.BaseAddress = new Uri("https://github.com/"));
```

Default configuration can be added to multiple times, between each other it will be applied one-by-one in order of registration; and its place in the registration compared to `AddHttpClient` doesn't matter:

```csharp
services.ConfigureHttpClientDefaults(b =>
    b.ConfigureHttpClient(c => c.BaseAddress = new Uri("https://consoto.com/")));

// Client a base address of https://consoto.com and has MyAuthHandler
services.AddHttpClient("consoto");

services.AddHttpClientDefaults()
    .AddHttpMessageHandler<MyAuthHandler>();
```

#### 3. Subsequent per-name reconfiguration

Setting additional properties in a primary handler:

```c#
services.ConfigureHttpClientDefaults(b =>
    b.ConfigurePrimaryHandler(() => new SocketsHttpHandler() { UseCookies = false }));

// will have SocketsHttpHandler with UseCookies = false and MaxConnectionsPerServer = 1
services.AddHttpClient("bar")
    .ConfigurePrimaryHandler((handler, _) => ((SocketsHttpHandler)handler).MaxConnectionsPerServer = 1);
```

Inserting in a specific position or removing from an additional handlers list:

```c#
services.ConfigureHttpClientDefaults(b =>
    b.AddHttpMessageHandler<MyAuthHandler>());

// will have MyLoggingHandler as a top-most wrapper handler, and MyAuthHandler
services.AddHttpClient("baz")
    .ConfigureAdditionalHttpMessageHandlers((handlerList, _) => handlerList.Insert(0, new MyLoggingHandler());

// will NOT have MyAuthHandler
services.AddHttpClient("qux")
    .ConfigureAdditionalHttpMessageHandlers((handlerList, _) =>
        handlerList.Remove(handlerList.SingleOrDefault(h => h.GetType() == typeof(MyAuthHandler))));
```

#### 4. Subsequent per-name reconfiguration -- order of applying

Between each other, the order of `ConfigurePrimaryHandler` and `ConfigureAdditionalHttpMessageHandlers` and other per-name configuration methods is the same as if it was implemented via `ConfigureHttpMessageHandlerBuilder` -- in order of registration, and if a subsequent write overwrites e.g. a primary handler instance, the last one "wins".

```c#
// Primary handler will already be a SocketsHttpHandler instance because it was set in ConfigureHttpClientDefaults
// below, which runs first, so it is safe to cast
services.AddHttpClient("bar")
    .ConfigurePrimaryHandler((handler, _) => ((SocketsHttpHandler)handler).MaxConnectionsPerServer = 1);

// But the second call overwrites primary handler, so the final result will NOT have MaxConnectionsPerServer  set
services.AddHttpClient("bar")
    .ConfigurePrimaryHandler(() => new SocketsHttpHandler() { UseCookies = false }));

services.ConfigureHttpClientDefaults(b =>
    b.ConfigurePrimaryHandler(() => new SocketsHttpHandler()));
```

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/runtime/issues/87914

The `AddHttpClientDefaults` method supports adding configuration to all created `HttpClients`.

The method:

* Creates a builder with a null name. `Microsoft.Extensions.Configuration` automatically applies configuration with a null name to all named configuration.
* Ensures that default configuration is added before named configuration in the `IServiceCollection`. This is to make it so the order of `AddHttpClientDefaults` and `AddHttpClient` doesn't matter. Default config is always applied first, then named config is applied after. This is done by wrapping the `IServiceCollection` in an implementation that modifies the order that `IConfigureOptions<HttpClientFactoryOptions>` values are added.

### CODE REVIEW

**Comment by davidfowl** [comment]
> Ensures that default configuration is added before named configuration in the IServiceCollection. This is to make it so the order of AddHttpClientDefaults and AddHttpClient doesn't matter. Default config is always applied first, then named config is applied after. This is done by wrapping the IServiceCollection in an implementation that modifies the order that IConfigureOptions<HttpClientFactoryOptions> values are added.

This feels pretty bad.

**Comment by JamesNK** [comment]
@noahfalk @joperezr @geeknoid @davidfowl 

The `AddHttpClientDefaults` is designed to make it easy to apply configuration to all clients configured with `AddHttpClient` using one extension method on `IHttpClientBuilder`.

The goal is to avoid having two methods:
* Add config to one builder
* Add config to all builders

For example, methods in dotnet/extensions below will no longer be needed:

https://github.com/dotnet/extensions/blob/52ebc1b123613254f9dfe250515027c1332fba35/src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/HttpClientLoggingExtensions.cs#L31-L122

**Comment by davidfowl** [comment]
I've also been thinking about an API like this and was hoping we could add an overload to AddHttpClient so it wasnt 2 different calls. 

services.AddHttpClient(IHttpClientBuilder clientBuilder =>
{
     clientBuilder.ConfigureHttpClient();
});

**Comment by JamesNK** [comment]
> > Ensures that default configuration is added before named configuration in the IServiceCollection. This is to make it so the order of AddHttpClientDefaults and AddHttpClient doesn't matter. Default config is always applied first, then named config is applied after. This is done by wrapping the IServiceCollection in an implementation that modifies the order that IConfigureOptions values are added.
> 
> This feels pretty bad.

It's not the best, but it works well. It hits these important requirements:

* `AddHttpClientDefaults` config is overridden by named client configuration.
* The order `AddHttpClientDefaults` and `AddHttpClient` are called doesn't matter. Named client configuration always wins.
* It automatically works for people who have written their own `IHttpClientBuilder` extension methods that call `services.Configure<HttpClientFactoryOptions>(...)`.

The extra logic is only applied to `IOptionsConfiguration<HttpClientFactoryOptions>` to ensure they have the right order. It won't interfere with or break other DI extension methods.

**Comment by geeknoid** [comment]
@dpk83 FYI

**Comment by dpk83** [comment]
@xakep139

File: src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        public void AddHttpClientDefaults_WithNameConfig_NameConfigUsed()
+        {
+            // Arrange
+            var serviceCollection = new ServiceCollection();
+
+            // Act1
+            serviceCollection.AddHttpClient("example.com", c => c.BaseAddress = new Uri("http://example.com/"));
+            serviceCollection.AddHttpClientDefaults().ConfigureHttpClient(c => c.BaseAddress = new Uri("http://default.com/"));
+
+            var services = serviceCollection.BuildServiceProvider();
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs:24]: using var?

Comment (JamesNK) [reply src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs:24]: Yup

**Comment by stephentoub** [comment]
Let's please agree on the API via discussion on the issue and API review before opening PRs.  I spoke with @CarnaViire at length about this and other related APIs yesterday and I know she was going to be following up with you.

File: src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs (hunk 1 above)
Comment (xakep139) [reply src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs:24]: I meant to dispose built service collection 🙂
`using var services = ...`

Comment (JamesNK) [reply src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DependencyInjection/AddHttpClientDefaultsTest.cs:24]: Ohhhh.

No, that's not necessary. The test isn't checking that disposable things are disposed of. It wouldn't hurt, but none of the other tests dispose the service provider.

**Review by CarnaViire — APPROVED** [review]
LGTM for @JamesNK 's changes

**Comment by CarnaViire** [comment]
@dotnet/ncl can someone pls review as well? Since I've taken over the PR.

**Review by ManickaP — APPROVED** [review]
@CarnaViire changes LGTM.

**Comment by lewing** [comment]
This change appears to be breaking flow into aspnetcore in https://github.com/dotnet/aspnetcore/pull/49433

**Comment by joperezr** [comment]
@CarnaViire can you please take a look?

**Comment by CarnaViire** [comment]
I've pushed the fix to the branch https://github.com/dotnet/aspnetcore/pull/49433

**Comment by davidfowl** [comment]
So we have a problem here. I tried to use **ConfigureHttpClientDefaults** and noticed it isn't an extension method in the reference assembly.

**Comment by joperezr** [comment]
Can you elaborate? This change seems to suggest that the reference assembly was edited and that the extension was properly added: https://github.com/dotnet/runtime/pull/87953/files#diff-06e422c9663ed696b723dc56832cbdb53b9f9c3e6d0a98e52a1cb081b7c9dc7aR50

**Comment by karelz** [comment]
@joperezr the `this` is missing -- see #90911 which David filed and which was just fixed :)

**Comment by joperezr** [comment]
Oh, completely missed it! Good catch @davidfowl and thanks for the quick fix 😃

**Pull request merged**